libherokubuildpack/src/buildpack_output/mod.rs

```diff
@@ -0,0 +1,500 @@
+                Self::style(format!("Done {details}")),
+            );
+        } else {
+            writeln_now(&mut self.state.write, Self::style("Done"));
+        }
+
+        self.state.write.inner
+    }
+}
+
+impl<W> BuildpackOutput<state::Section<W>>
+where
+    W: Write + Send + Sync + 'static,
+{
+    const PREFIX_FIRST: &'static str = "  - ";
+    const PREFIX_REST: &'static str = "    ";
+    const CMD_INDENT: &'static str = "      ";
+
+    fn style(s: impl AsRef<str>) -> String {
+        prefix_first_rest_lines(Self::PREFIX_FIRST, Self::PREFIX_REST, s.as_ref())
+    }
+
+    #[must_use]
+    pub fn step(mut self, s: impl AsRef<str>) -> BuildpackOutput<state::Section<W>> {
+        writeln_now(&mut self.state.write, Self::style(s));
+        self
+    }
+
+    pub fn start_stream(mut self, s: impl AsRef<str>) -> BuildpackOutput<state::Stream<W>> {
+        writeln_now(&mut self.state.write, Self::style(s));
+        writeln_now(&mut self.state.write, "");
+
+        BuildpackOutput {
+            started: self.started,
+            state: state::Stream {
+                started: Instant::now(),
+                write: line_mapped(self.state.write, |mut input| {
+                    if input.iter().all(u8::is_ascii_whitespace) {
+                        input
+                    } else {
+                        let mut result: Vec<u8> = Self::CMD_INDENT.into();
+                        result.append(&mut input);
+                        result
+                    }
+                }),
+            },
+        }
+    }
+
+    pub fn finish(self) -> BuildpackOutput<state::Started<W>> {
+        BuildpackOutput {
+            started: self.started,
+            state: state::Started {
+                write: self.state.write,
+            },
+        }
+    }
+}
+
+impl<W> BuildpackOutput<state::Stream<W>>
+where
+    W: Write + Send + Sync + 'static,
+{
+    pub fn finish(mut self) -> BuildpackOutput<state::Section<W>> {
+        let duration = self.state.started.elapsed();
+
+        writeln_now(&mut self.state.write, "");
+
+        BuildpackOutput {
+            started: self.started,
+            state: state::Section {
+                write: self.state.write.unwrap(),
+            },
+        }
+        .step(format!(
+            "Done {}",
+            style::details(duration_format::human(&duration))
+        ))
+    }
+}
+
+impl<W> Write for BuildpackOutput<state::Stream<W>>
+where
+    W: Write,
+{
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.state.write.write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.state.write.flush()
+    }
+}
+
+/// Internal helper, ensures that all contents are always flushed (never buffered).
+fn writeln_now<D: Write>(destination: &mut D, msg: impl AsRef<str>) {
+    writeln!(destination, "{}", msg.as_ref()).expect("Output error: UI writer closed");
+
+    destination.flush().expect("Output error: UI writer closed");
```

Comment:
(nit) The error message style used for these `.expect()`s don't match the guidelines here:
https://doc.rust-lang.org/std/error/index.html#common-message-styles